crates/goose/src/agent.rs

```diff
@@ -161,6 +162,99 @@ impl Agent {
         Ok(status)
     }
 
+    /// Gets token counts, resource metadata and content for each resource URI across all systems
+    async fn get_system_status_counts(&self) -> AgentResult<HashMap<String, HashMap<String, (Resource, u32, String)>>> {
+        let mut system_token_counts = HashMap::new();
+        let token_counter = TokenCounter::new();
+    
+        for system in &self.systems {
+            let system_status = system
+                .status()
+                .await
+                .map_err(|e| AgentError::Internal(e.to_string()))?;
+    
+            let mut resource_counts = HashMap::new();
+            for resource in system_status {
+                if let Ok(content) = system.read_resource(&resource.uri).await {
+                    let token_count = token_counter.count_tokens(&content, None) as u32;
+                    resource_counts.insert(resource.uri.clone(), (resource, token_count, content));
+                }
+            }
+    
+            system_token_counts.insert(system.name().to_string(), resource_counts);
+        }
+        
+        Ok(system_token_counts)
+    }
+
+
+    /// Trims the system status to reduce size while preserving critical information by sorting resources
+    /// by priority and timestamp, removing older and lower priority resources to meet token limits
+    fn trim_system_status(&self, status_counts: HashMap<String, HashMap<String, (Resource, u32, String)>>, approx_count: usize, context_limit: usize) -> AgentResult<String> {
+        // Flatten all resources into a single vector for sorting
+        let mut all_resources: Vec<(String, String, Resource, u32, String)> = Vec::new();
+        for (system_name, resources) in status_counts.iter() {
+            for (uri, (resource, token_count, content)) in resources {
+                all_resources.push((
+                    system_name.clone(),
+                    uri.clone(),
+                    resource.clone(),
+                    *token_count,
+                    content.clone(),
+                ));
+            }
+        }
+
+        // Sort by priority (high to low) and timestamp (newest to oldest)
```

Comment:
is priority a discrete value or float between 0-1? if its a float, unlikely we'll have many items in the secondary time sort

Reply:
it's a float to be consistent with the priority on content. That's a good point, but I think that's OK for now. Currently, we only have 0's and 1's for priority

Reply:
yeah +1 that this is okay, the prevalance of zeros is what makes us need a tiebreaker. Will revisit this over time i'm sure